trait bound in rpc attibute marco is ignored

#### Problem description
I run into this issue when I was trying to build a custom RPC in substrate whose return value is a `BTreeMap<AccountId, _>`, in which case `Ord` is required for `AccountId`.

#### To reproduce

It's easy to reproduce using the example in jsonrpc-derive crate with a bit of alternation.

```diff
diff --git a/derive/examples/generic-trait.rs b/derive/examples/generic-trait.rs
index 6cf867b..75bdbdb 100644
--- a/derive/examples/generic-trait.rs
+++ b/derive/examples/generic-trait.rs
@@ -3,16 +3,20 @@ use jsonrpc_core;
 use jsonrpc_core::futures::future::{self, FutureResult};
 use jsonrpc_core::{Error, IoHandler, Result};
 use jsonrpc_derive::rpc;
+use std::collections::BTreeMap;

 #[rpc]
-pub trait Rpc<One, Two> {
+pub trait Rpc<One, Two>
+where
+       One: Ord,
+{
        /// Get One type.
        #[rpc(name = "getOne")]
        fn one(&self) -> Result<One>;

        /// Adds two numbers and returns a result
        #[rpc(name = "setTwo")]
-       fn set_two(&self, a: Two) -> Result<()>;
+       fn set_two(&self, a: Two) -> Result<BTreeMap<One, ()>>;

        /// Performs asynchronous operation
        #[rpc(name = "beFancy")]
@@ -26,9 +30,9 @@ impl Rpc<u64, String> for RpcImpl {
                Ok(100)
        }

-       fn set_two(&self, x: String) -> Result<()> {
+       fn set_two(&self, x: String) -> Result<BTreeMap<u64, ()>> {
                println!("{}", x);
-               Ok(())
+               Ok(Default::default())
        }

        fn call(&self, num: u64) -> FutureResult<(u64, String), Error> {
```

```
xlc@xlc-pc:~/.../jsonrpc/derive ❯❯❯ cargo run --example generic-trait
   Compiling jsonrpc-derive v14.2.1 (/home/xlc/data/src/github.com/paritytech/jsonrpc/derive)
error[E0277]: the trait bound `One: std::cmp::Ord` is not satisfied
 --> derive/examples/generic-trait.rs:8:1
  |
8 | #[rpc]
  | ^^^^^^ the trait `std::cmp::Ord` is not implemented for `One`
  |
  = note: required because of the requirements on the impl of `for<'de> jsonrpc_core::serde::Deserialize<'de>` for `std::collections::BTreeMap<One, ()>`
  = note: required because of the requirements on the impl of `jsonrpc_core::serde::de::DeserializeOwned` for `std::collections::BTreeMap<One, ()>`
  = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider further restricting this bound
  |
8 | #[rpc] + std::cmp::Ord
  |        ^^^^^^^^^^^^^^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
error: could not compile `jsonrpc-derive`.

To learn more, run the command again with --verbose.
```

I believe this issue is caused by the `gen_client` part in this macro.

```rust
    /// The generated client module.
    pub mod gen_client {
......
    /// `Ord` bound is ignored in the expanded code unexpectedly.
        impl<One, Two> Client<One, Two>
        where
            One: Send
                + Sync
                + 'static
                + _jsonrpc_core::serde::de::DeserializeOwned
                + _jsonrpc_core::serde::Serialize,
```

--- a/derive/src/to_delegate.rs
+++ b/derive/src/to_delegate.rs
@@ -7,7 +7,7 @@ use syn::{
 	parse_quote,
 	punctuated::Punctuated,
 	visit::{self, Visit},
-	Result, Token,
+	Result, Token, WherePredicate,
 };
 
 pub enum MethodRegistration {
@@ -458,6 +458,8 @@ pub fn generate_where_clause_serialization_predicates(
 	let mut visitor = FindTyParams::default();
 	visitor.visit_item_trait(item_trait);
 
+	let additional_where_clause = item_trait.generics.where_clause.clone();
+
 	item_trait
 		.generics
 		.type_params()
@@ -483,6 +485,20 @@ pub fn generate_where_clause_serialization_predicates(
 					bounds.push(parse_quote!(_jsonrpc_core::serde::de::DeserializeOwned))
 				}
 			}
+
+			// add the trait bounds specified by the user in where clause.
+			if let Some(ref where_clause) = additional_where_clause {
+				for predicate in where_clause.predicates.iter() {
+					if let WherePredicate::Type(where_ty) = predicate {
+						if let syn::Type::Path(ref predicate) = where_ty.bounded_ty {
+							if *predicate == ty_path {
+								bounds.extend(where_ty.bounds.clone().into_iter());
+							}
+						}
+					}
+				}
+			}
+
 			syn::WherePredicate::Type(syn::PredicateType {
 				lifetimes: None,
 				bounded_ty: syn::Type::Path(ty_path),

--- a/derive/src/to_delegate.rs
+++ b/derive/src/to_delegate.rs
@@ -7,7 +7,7 @@ use syn::{
 	parse_quote,
 	punctuated::Punctuated,
 	visit::{self, Visit},
-	Result, Token, WherePredicate,
+	Result, Token,
 };
 
 pub enum MethodRegistration {
@@ -489,7 +489,7 @@ pub fn generate_where_clause_serialization_predicates(
 			// add the trait bounds specified by the user in where clause.
 			if let Some(ref where_clause) = additional_where_clause {
 				for predicate in where_clause.predicates.iter() {
-					if let WherePredicate::Type(where_ty) = predicate {
+					if let syn::WherePredicate::Type(where_ty) = predicate {
 						if let syn::Type::Path(ref predicate) = where_ty.bounded_ty {
 							if *predicate == ty_path {
 								bounds.extend(where_ty.bounds.clone().into_iter());

--- a/derive/tests/run-pass/client_with_generic_trait_bounds.rs
+++ b/derive/tests/run-pass/client_with_generic_trait_bounds.rs
@@ -1,0 +1,39 @@
+use jsonrpc_core::futures::future::{self, FutureResult};
+use jsonrpc_core::{Error, IoHandler, Result};
+use jsonrpc_derive::rpc;
+use std::collections::BTreeMap;
+
+#[rpc]
+pub trait Rpc<One, Two>
+where
+	One: Ord,
+	Two: Ord + Eq,
+{
+	/// Adds two numbers and returns a result
+	#[rpc(name = "setTwo")]
+	fn set_two(&self, a: Two) -> Result<BTreeMap<One, ()>>;
+
+	/// Performs asynchronous operation
+	#[rpc(name = "beFancy")]
+	fn call(&self, a: One) -> FutureResult<(One, Two), Error>;
+}
+
+struct RpcImpl;
+
+impl Rpc<u64, String> for RpcImpl {
+	fn set_two(&self, x: String) -> Result<BTreeMap<u64, ()>> {
+		println!("{}", x);
+		Ok(Default::default())
+	}
+
+	fn call(&self, num: u64) -> FutureResult<(u64, String), Error> {
+		crate::future::finished((num + 999, "hello".into()))
+	}
+}
+
+fn main() {
+	let mut io = IoHandler::new();
+	let rpc = RpcImpl;
+
+	io.extend_with(rpc.to_delegate())
+}

